sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/AnsiTypeCoercion.scala

```diff
@@ -148,6 +149,10 @@ object AnsiTypeCoercion extends TypeCoercionBase {
       // interval type the string should be promoted as. There are many possible interval
       // types, such as year interval, month interval, day interval, hour interval, etc.
       case (_: StringType, _: AnsiIntervalType) => None
+      // [SPARK-50060] If a binary operation contains at least one collated string types, we can't
+      // decide which collation the result should have.
+      case (d1: StringType, d2: StringType) if SchemaUtils.hasNonUTF8BinaryCollation(d1) ||
```

Comment:
This makes me think that maybe we should not put string collation type coercion in an individual rule, as other type coercion rules must be aware of string collation as well. Shall we just make each type coercion rule collation aware?

Reply:
After adding the case when we have collated strings as column types in the output of set operators, we do not have to modify the `AnsiTypeCoercion` and `TypeCoercion` and add this case-matchings.